core/src/main/java/org/elasticsearch/index/query/TermsQueryParser.java

```diff
@@ -159,58 +140,18 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
         }
 
         if (fieldName == null) {
-            throw new QueryParsingException(parseContext, "terms query requires a field name, followed by array of terms");
-        }
-
-        MappedFieldType fieldType = context.fieldMapper(fieldName);
-        if (fieldType != null) {
-            fieldName = fieldType.names().indexName();
-        }
-
-        if (lookupId != null) {
-            final TermsLookup lookup = new TermsLookup(lookupIndex, lookupType, lookupId, lookupRouting, lookupPath, parseContext);
-            GetRequest getRequest = new GetRequest(lookup.getIndex(), lookup.getType(), lookup.getId()).preference("_local").routing(lookup.getRouting());
-            getRequest.copyContextAndHeadersFrom(SearchContext.current());
-            final GetResponse getResponse = client.get(getRequest).actionGet();
-            if (getResponse.isExists()) {
-                List<Object> values = XContentMapValues.extractRawValues(lookup.getPath(), getResponse.getSourceAsMap());
-                terms.addAll(values);
-            }
-        }
-
-        if (terms.isEmpty()) {
-            return Queries.newMatchNoDocsQuery();
-        }
-
-        Query query;
-        if (context.isFilter()) {
-            if (fieldType != null) {
-                query = fieldType.termsQuery(terms, context);
-            } else {
-                BytesRef[] filterValues = new BytesRef[terms.size()];
-                for (int i = 0; i < filterValues.length; i++) {
-                    filterValues[i] = BytesRefs.toBytesRef(terms.get(i));
-                }
-                query = new TermsQuery(fieldName, filterValues);
-            }
-        } else {
-            BooleanQuery bq = new BooleanQuery(disableCoord);
-            for (Object term : terms) {
-                if (fieldType != null) {
-                    bq.add(fieldType.termQuery(term, context), Occur.SHOULD);
-                } else {
-                    bq.add(new TermQuery(new Term(fieldName, BytesRefs.toBytesRef(term))), Occur.SHOULD);
-                }
-            }
-            Queries.applyMinimumShouldMatch(bq, minShouldMatch);
-            query = bq;
-        }
-        query.setBoost(boost);
-
-        if (queryName != null) {
-            context.addNamedQuery(queryName, query);
+            throw new QueryParsingException(parseContext, "terms query requires a field name, followed by array of terms or a document lookup specification");
         }
-        return query;
+        return new TermsQueryBuilder(fieldName, values)
+                .disableCoord(disableCoord)
+                .minimumShouldMatch(minShouldMatch)
+                .lookupIndex(lookupIndex)
+                .lookupType(lookupType)
+                .lookupId(lookupId)
+                .lookupRouting(lookupRouting)
+                .lookupPath(lookupPath)
```

Comment:
As mentioned above, I'd opt for setting the fully constructed lookUp oject here in case it is valid.